simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/actions/Authorize.java

```diff
@@ -40,49 +42,60 @@ public void perform(StationState stationState, StationMessageSender stationMessa
 
         stationMessageSender.sendAuthorizeAndSubscribe(tokenId, evseIds, (request, response) -> {
             if (response.getIdTokenInfo().getStatus() == IdTokenInfo.Status.ACCEPTED) {
-                List<Integer> authorizedEvseIds = response.getEvseId() == null || response.getEvseId().isEmpty() ? singletonList(stationState.getDefaultEvseId()) : response.getEvseId();
-                authorizedEvseIds.forEach(evseId -> stationState.storeToken(evseId, tokenId));
+                List<Evse> authorizedEvses = hasEvses(response) ? getEvseList(response, stationState) : singletonList(stationState.getDefaultEvse());
 
-                boolean haveOngoingTransaction = authorizedEvseIds.stream().allMatch(stationState::hasOngoingTransaction);
+                authorizedEvses.forEach(evse -> evse.setToken(tokenId));
+
+                boolean haveOngoingTransaction = authorizedEvses.stream().allMatch(Evse::hasOngoingTransaction);
 
                 if (!haveOngoingTransaction) {
                     Integer transactionId = TransactionIdGenerator.getInstance().getAndIncrement();
-                    authorizedEvseIds.forEach(evseId -> stationState.findEvse(evseId).createTransaction(transactionId));
+                    authorizedEvses.forEach(evse -> evse.createTransaction(transactionId));
                 }
 
-                boolean allCharging = authorizedEvseIds.stream().allMatch(stationState::isCharging);
-                boolean allPlugged = authorizedEvseIds.stream().allMatch(stationState::isPlugged);
+                boolean allCharging = authorizedEvses.stream().allMatch(Evse::isCharging);
+                boolean allPlugged = authorizedEvses.stream().allMatch(Evse::isCablePlugged);
 
-                if (allPlugged) {
-                    startCharging(stationState, stationMessageSender, authorizedEvseIds);
+                if (allPlugged && !allCharging) {
```

Comment:
```
if (allCharging) {
    stopCharging(stationMessageSender, authorizedEvses);
} else { // !allCharging
    if (allPlugged) {
        startCharging(stationMessageSender, authorizedEvses);
    } else { // !allPlugged && !allCharging
        if (haveOngoingTransaction) {
            startCharging(stationMessageSender, authorizedEvses);
        } else { // !allPlugged && !allCharging && !haveOngoingTransaction
            authorizedEvses.forEach(evse -> stationMessageSender.sendTransactionEventStart(evse.getId(), AUTHORIZED, tokenId));
        }
    }
}
```

^ is this the same as your condition? If so I think it's a lot more readable :)

Reply:
Cool. Thanks

Reply:
Just please double check that it's actually the same condition, I didn't spend a lot of time checking it.

Reply:
Sure. It is cover with unit tests